bleak/__init__.py

```diff
@@ -204,6 +205,51 @@ def set_scanning_filter(self, **kwargs):
         )
         self._backend.set_scanning_filter(**kwargs)
 
+    async def devices(self) -> AsyncIterator[Tuple[BLEDevice, AdvertisementData]]:
+        """
+        Yields devices and their advertising data as they are discovered.
+
+        .. note::
+            Ensure that scanning is started before calling this method.
+
+        Returns:
+            An async iterator that yields tuples (:class:`BLEDevice`, :class:`AdvertisementData`).
+        """
+        if self._backend.detection_callback:
+            raise BleakError(
+                "Cannot use async iterator methods when detection_callback is used"
+            )
+
+        devices = asyncio.Queue()
+
+        async def detection_callback(bd, ad):
+            await devices.put((bd, ad))
+
+        self._backend.register_detection_callback(detection_callback)
+
+        while True:
+            try:
+                yield await devices.get()
+            except GeneratorExit:
```

Comment:
Fix: I will change `except GeneratorExit:` to `finally:` on next force push

Reply:
Technically, this is going to suffer the same leak when `devices()` is not iterated to completion as mentioned in my other comments. However, in this case it doesn't seem like the consequences are that bad since we usually leave the callback hanging anyway. The only problem would be if someone called `devices()` a second time without waiting causing it to raise the `BleakError` because of the lingering callback, but this can be worked around.

I suppose we could also consider just leaving out the `try/finally` and the error check.

Reply:
> Fix: I will change `except GeneratorExit:` to `finally:` on next force push

Note to self: just changing it raises https://rules.sonarsource.com/python/RSPEC-1143/ . However, I solved it by moving the `while True`.

> ...  this is going to suffer the same leak when devices() is not iterated to completion ...

As for the callback, I would argue that if this method registers it, it shall also unregister it for sure. Lingering callback would throw error anyway because `devices` queue may/would not be allocated anymore.

For the queue, I have thought about the queue cleaning before adding it (even in the first implementation), but somewhere found something in lines of
> When you set a variable to None, the reference to the object is lost. Python's garbage collector will eventually deallocate the object when there are no more references to it - including items left in the queue, assuming there are no other references to these objects.

However, as I cannot find this again, neither any actual source of it, I have added explicit cleanup. Still not sure whether the explicit cleanup is required or `devices = None` would do :thinking: 
Also note the added comment to commit message about why the queue is not cleaned using 
```python
try:
    while True:
        devices.get_nowait()
except asyncio.QueueEmpty:
    pass
```

Reply:
Explicit queue cleanup removed, as per https://github.com/hbldh/bleak/pull/1361#discussion_r1263876425